librz/analysis/arch/x86/il_fp_ops.inc

```diff
@@ -0,0 +1,697 @@
+		UGE(VARL("i"), UN(mem_size, 0)), // i >= 0
+		SEQ3(
+			set_byte, // get byte
+			SETL("val", new_val), // update val
+			SETL("i", SUB(VARL("i"), UN(mem_size, 1))) // i--
+			));
+
+	RzILOpEffect *f_init = SETL("f", x86_il_floating_from_int(VARL("val"), RZ_FLOAT_IEEE754_BIN_80));
+
+	/* Check sign byte (index 9) checking if sign byte is zero */
+	RzILOpPure *sign_byte = LOADW(8, ADD(VARL("mem"), UN(mem_size, 9)));
+	RzILOpFloat *final_float = ITE(IS_ZERO(sign_byte), VARL("f"), FNEG(VARL("f")));
+
+	return SEQ6(
+		mem_init, i_init, val_init, // Local vars init
+		bcd_decode_loop, // BCD decoding loop
+		f_init, // Conversion to 80-bit float
+		x86_il_st_push(final_float, RZ_FLOAT_IEEE754_BIN_80) // Push the value onto the FPU stack
+	);
+}
+
+/**
+ * FBSTP
+ * Pop the float in ST(0), convert it to BCD and store
+ */
+IL_LIFTER(fbstp) {
+	ut8 mem_size = analysis->bits;
+
+	RzILOpEffect *pop_eff;
+	RzILOpPure *pop_val;
+	X86_IL_ST_POP(pop_val, pop_eff);
+
+	RzILOpEffect *val_init = SETL("val", x86_il_int_from_floating(pop_val, 64));
+	RzILOpEffect *sgn_init = SETL("sgn", MSB(VARL("val")));
+	RzILOpEffect *val_abs = SETL("val", ITE(VARL("sgn"), NEG(VARL("val")), VARL("val")));
+
+	RzILOpEffect *mem_init = SETL("mem", x86_il_get_memaddr(ins->structure->operands[0].mem));
+	RzILOpEffect *i_init = SETL("i", UN(mem_size, 0));
+	RzILOpPure *byte_mem = ADD(VARL("mem"), VARL("i"));
+
+	RzILOpPure *upper_nibble = DIV(MOD(VARL("val"), UN(64, 100)), UN(64, 10));
+	RzILOpPure *lower_nibble = MOD(VARL("val"), UN(64, 10));
+	RzILOpPure *new_byte = UNSIGNED(8, LOGOR(SHIFTL0(upper_nibble, UN(8, 4)), lower_nibble));
+
+	RzILOpEffect *bcd_encode_loop = REPEAT( // while
+		ULT(VARL("i"), UN(mem_size, 9)), // i < 9
+		SEQ3(
+			STOREW(byte_mem, new_byte), // store byte
+			SETL("val", DIV(VARL("val"), UN(64, 100))), // val /= 100
+			SETL("i", ADD(VARL("i"), UN(mem_size, 1))) // i++
+			));
+
+	RzILOpEffect *update_mem = SETL("mem", DUP(byte_mem));
+	RzILOpEffect *store_sign = BRANCH(VARL("sgn"), STOREW(VARL("mem"), UN(8, 0xff)), STOREW(VARL("mem"), UN(8, 0)));
+
+	RzILOpEffect *noexcept_case = SEQ5(
+		mem_init, i_init, // init loop vars
+		bcd_encode_loop, // convert val to bcd
+		update_mem, store_sign // update mem pointer and store the saved sign in the bcd
+	);
+	RzILOpEffect *maybe_except = BRANCH(
+		SGE(VARL("val"), UN(64, 1e18)), // |val| >= 1e18
+		GOTO("int"), // throw exception if val can't fit in 18-digit BCD
+		noexcept_case);
+
+	return SEQ5(
+		val_init, pop_eff, // get val from ST(0) and pop
+		sgn_init, val_abs, // val = |val| and store sign
+		maybe_except);
+}
+
+/* Arithmetic instructions */
+
+/**
+ * FABS
+ * Clears the sign bit of ST(0) to create absolute value
+ */
+IL_LIFTER(fabs) {
+	RzILOpFloat *abs_value = FABS(x86_il_get_st_reg(X86_REG_ST0));
+	return SEQ2(x86_il_set_st_reg(X86_REG_ST0, abs_value, RZ_FLOAT_IEEE754_BIN_80), x86_il_set_fpu_flag(X86_FPU_C1, IL_FALSE));
+}
+
+#define FLOATING_ARITHMETIC_IL(op) \
```

Comment:
More a decision of taste. But is there a reason this is a macro and not a function?
Those macros can't be debugged so it could be a little annoying in the future.

Reply:
We need to pass in different ops, of which each have their own `x86_il_##op##_with_rmode` function. I could probably have a macro which converts the op to the function name and passes that function pointer into a normal function, which can then call that function pointer. That would reduce the code in the macro, but would be (slightly) more convoluted. I would prefer it though.

Reply:
It's fine for me then.